### PR TITLE
[#146923663] idp authentication failure

### PR DESCRIPTION
**Pull request by scerreto**

### CODE REVIEW

File: js/components/SpidLoginButton.js [hunk 1]
```diff
@@ -441,5 +454,12 @@ const styles = StyleSheet.create({
   spinner: {
     width: 20,
     marginRight: 6
+  },
+  errorMissingToken: {
+    padding: 5,
+    backgroundColor: '#F23333',
```
Comment (cloudify) [inline js/components/SpidLoginButton.js]: are we going to refactor color codes somewhere in a single place? or are we going to wait the moment when the designer says "I'd like to change that color everywhere"? :D

Comment (gunzip) [reply js/components/SpidLoginButton.js]: Maybe we don't have to deal with styles at all in this development phase, just use defaults ?

File: js/reducers/user.js [hunk 2]
```diff
@@ -21,6 +22,7 @@ import config from '../config'
 // user state when logged out
 export type LoggedOutUserState = {
   isLoggedIn: false,
+  loginError?: boolean,
```
Comment (cloudify) [inline js/reducers/user.js]: boolean attributes/variables should be named `isSomething` or `hasSomething`, in this case `isError`

said that, I'm not sure what it does mean, please add a short comment describing what it means when that attribute is `true` - also, perhaps, worth considering splitting the `LoggedOutUserState` type in two: `DefaultLoggedOutUserState` and `FailedLoggedOutUserState`

File: js/reducers/user.js [hunk 3]
```diff
@@ -39,6 +41,7 @@ export type UserState = LoggedOutUserState | LoggedInUserState
 // initial user state
 const initialUserState: LoggedOutUserState = {
```
Comment (cloudify) [inline js/reducers/user.js]: this would be a `DefaultLoggedOutUserState`

File: js/reducers/user.js [hunk 4]
```diff
@@ -79,5 +83,11 @@ export default function user(
     }
   }
 
+  if (action.type === USER_LOGIN_ERROR_ACTION && state.isLoggedIn === false) {
+    return {
+      loginError: true
```
Comment (cloudify) [inline js/reducers/user.js]: this doesn't look like a valid `UserState`

File: js/components/SpidLoginButton.js (hunk 1 above)
Comment (scerreto) [reply js/components/SpidLoginButton.js]: waiting the final graphic, i add it in common style (components/style,js) (errorContent)

File: js/reducers/user.js (hunk 4 above)
Comment (scerreto) [reply js/reducers/user.js]: fixed

File: js/reducers/user.js (hunk 3 above)
Comment (scerreto) [reply js/reducers/user.js]: fixed

File: js/reducers/user.js (hunk 2 above)
Comment (scerreto) [reply js/reducers/user.js]: fixed

File: js/reducers/user.js [hunk 5]
```diff
@@ -20,9 +20,15 @@ import type { ApiUserProfile } from '../utils/api'
...
 // user state when logged out
-export type LoggedOutUserState = {
+export type DefaultLoggedOutUserState = {
   isLoggedIn: false,
-  loginError?: boolean,
+  apiUrlPrefix: string
+}
+
+// user state when there is an error, tracing by isError=true
+export type ErrorLoggedOutUserState = {
```
Comment (cloudify) [inline js/reducers/user.js]: to make a [disjoint union](https://flow.org/en/docs/types/unions/#toc-disjoint-unions) work you need a discriminating attribute, in this case the discriminator between `DefaultLoggedOutUserState` and `ErrorLoggedOutUserState` is the value of `isError` - what does it mean? it means that Flow, to be able to discriminate between `DefaultLoggedOutUserState` and `ErrorLoggedOutUserState` needs to know which is which based on the value of `isError`, and you tell flow by defining types with actual values, e.g.:

``` js
type DefaultLoggedOutUserState = {
  isError: false
}

type ErrorLoggedOutUserState = {
  isError: true
}
```

File: js/reducers/user.js [hunk 6]
```diff
@@ -85,7 +93,9 @@ export default function user(
 
   if (action.type === USER_LOGIN_ERROR_ACTION && state.isLoggedIn === false) {
     return {
-      loginError: true
+      isLoggedIn: false,
```
Comment (cloudify) [inline js/reducers/user.js]: @fpersico I wonder why in the previous PR the build didn't fail, do we have a step that checks the code with Flow?

Comment (cloudify) [reply js/reducers/user.js]: actually it did fail, I think we need a more verbose output in the build to see what failed (`--info` should be sufficient)

File: js/reducers/user.js (hunk 5 above)
Comment (scerreto) [reply js/reducers/user.js]: Fixed.

Comment (cloudify) [reply js/reducers/user.js]: how is it fixed? is the following a valid state/type?

``` js
type ErrorLoggedOutUserState = {
  isError: false
}
```

@fpersico @lussoluca what do you think?

Comment (fpersico) [reply js/reducers/user.js]: The updated code is
`
// user state when logged out
export type DefaultLoggedOutUserState = {
  isLoggedIn: false,
  apiUrlPrefix: string
}

// user state when there is an error, tracing by isError=true
export type ErrorLoggedOutUserState = {
  isLoggedIn: false,
  isError: true,
  apiUrlPrefix: string
}
`

DefaultLoggedOutUserState still need the `isError: false` to make disjoint union work

Comment (scerreto) [reply js/reducers/user.js]: Fixed DefaultLoggedOutUserState with isError: false

**Comment by cloudify** [comment]
please fix conflicts and rebase with `master`

File: js/components/SpidLoginButton.js [hunk 7]
```diff
@@ -239,6 +242,18 @@ class IdpSelectionScreen extends React.Component {
     })
   }
 
+  createErrorMessage = () => {
+    if (this.props.userState.isError === true) {
```
Comment (cloudify) [inline js/components/SpidLoginButton.js]: this check should happen at the caller, this function is named `createErrorMessage` and not `createErrorMessageOnlyIfUserStateIsError`

File: js/components/SpidLoginButton.js [hunk 8]
```diff
@@ -279,6 +294,7 @@ class IdpSelectionScreen extends React.Component {
           />
         ) : (
           <Content style={StyleSheet.flatten(styles.selectIdpContainer)}>
+            {this.createErrorMessage()}
```
Comment (cloudify) [inline js/components/SpidLoginButton.js]: the check on `this.props.userState.isError` should happen here

File: js/components/SpidLoginButton.js [hunk 9]
```diff
@@ -178,7 +180,8 @@ class IdpSelectionScreen extends React.Component {
     closeModal: () => void,
     onSelectIdp: IdentityProvider => void,
     onSpidLogin: (string, string) => void,
-    onSpidLoginError: string => void
+    onSpidLoginError: string => void,
+    userState: any
```
Comment (cloudify) [inline js/components/SpidLoginButton.js]: do not use `any`, we do have a type for [UserState](https://github.com/teamdigitale/italia-app/blob/master/js/reducers/user.js#L37)

File: js/components/SpidLoginButton.js [hunk 10]
```diff
@@ -322,7 +333,8 @@ export class SpidLoginButton extends React.Component {
     onSpidLoginIntent: () => void,
     onSelectIdp: IdentityProvider => void,
     onSpidLogin: (string, string) => void,
-    onSpidLoginError: string => void
+    onSpidLoginError: string => void,
+    userState: any
```
Comment (cloudify) [inline js/components/SpidLoginButton.js]: do not use `any`, we do have a type for [UserState](https://github.com/teamdigitale/italia-app/blob/master/js/reducers/user.js#L37)

**Pull request merged**